### DISCUSSION TITLE
AltSoftSerial

### DISCUSSION THREAD

**Discussion by Beelzebot42** [gpstar81/GPStar-proton-pack#61] [General]
Hi everyone. Thanks for letting me be a part. For context I am a useful idiot in this area. I can't really program, but I can assemble parts and solder them together, compile and upload code, ect...

I was running into an issue last night trying to compile the pack code. Was running into an error which I think was related to the AltSoftSerial library. I had it installed, but not added to the project. It seemed like adding a call  "#include <AltSoftSerial.h>" to my code fixed it. 

Is this expected? If it is I might suggest adding to the documentation. (I couldn't find it anywhere myself, but maybe I just missed it). 

I have so far soldered the pack project board together. It took a bit to decipher the schematic, but ultimately it made sense to me. Hoping to get my Mega setup and hooked up this evening to give the pack a test run.

Thanks again!

**Comment by gpstar81** [comment]
The pack does not need AltSoftSerial.

In the pack code, and on the Arduino instruction page here, you need to enable __WT_USE_SERIAL_3__ in the wavTrigger.h file.

More info can be found here: For compiling code on the Arduino Mega:

https://github.com/gpstar81/haslab-proton-pack/blob/main/ARDUINO.md#for-compiling-code-on-the-arduino-mega

**Comment by Beelzebot42** [comment]
When i remove the AltSoftSerial library for the pack and enable #define __WT_USE_SERIAL3__  I get the following errors when compiling. Orignially I thought adding AltSoftSerial fixed it, as it compiled without issue but the Mega doesn't seem to communicate with the WavTrigger when I upload that code. Everything else works, but no audio. The test function on the wavtrigger plays the first track. Any suggestions as to what I have configured wrong?


C:\Users\ruper\AppData\Local\Temp\ccps5E3X.ltrans0.ltrans.o: In function `update':
c:\arduino\libraries\libraries\WAV-Trigger-Arduino-Serial-Library-master/wavTrigger.cpp:67: undefined reference to `AltSoftSerial::available()'
c:\arduino\libraries\libraries\WAV-Trigger-Arduino-Serial-Library-master/wavTrigger.cpp:68: undefined reference to `AltSoftSerial::read()'
C:\Users\ruper\AppData\Local\Temp\ccps5E3X.ltrans0.ltrans.o: In function `flush':
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:55: undefined reference to `AltSoftSerial::flushOutput()'
C:\Users\ruper\AppData\Local\Temp\ccps5E3X.ltrans0.ltrans.o: In function `write':
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:54: undefined reference to `AltSoftSerial::writeByte(unsigned char)'
C:\Users\ruper\AppData\Local\Temp\ccps5E3X.ltrans0.ltrans.o: In function `global constructors keyed to 65535_0_proton_pack.ino.cpp.o.4858':
c:\arduino\libraries\libraries\ezButton\src/ezButton.cpp:34: undefined reference to `vtable for AltSoftSerial'
c:\arduino\libraries\libraries\ezButton\src/ezButton.cpp:34: undefined reference to `vtable for AltSoftSerial'
C:\Users\ruper\AppData\Local\Temp\ccps5E3X.ltrans0.ltrans.o: In function `AltSoftSerial::begin(unsigned long)':
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:47: undefined reference to `AltSoftSerial::init(unsigned long)'
C:\Users\ruper\AppData\Local\Temp\ccps5E3X.ltrans0.ltrans.o: In function `flush':
c:\arduino\libraries\libraries\WAV-Trigger-Arduino-Serial-Library-master/wavTrigger.cpp:53: undefined reference to `AltSoftSerial::available()'
c:\arduino\libraries\libraries\WAV-Trigger-Arduino-Serial-Library-master/wavTrigger.cpp:54: undefined reference to `AltSoftSerial::read()'
C:\Users\ruper\AppData\Local\Temp\ccps5E3X.ltrans0.ltrans.o: In function `AltSoftSerial::~AltSoftSerial()':
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:46: undefined reference to `vtable for AltSoftSerial'
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:46: undefined reference to `vtable for AltSoftSerial'
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:46: undefined reference to `AltSoftSerial::end()'
collect2.exe: error: ld returned 1 exit status

**Reply by Beelzebot42** [reply]
Also I have tried reinstalling my arduino IDE SW and all libraries fresh, and tried a different computer with fresh install. I am replicating these errors under each condition. Hopefully someone can tell me what I am doing wrong here?

**Comment by Beelzebot42** [comment]
I am sure I am just overlooking something here as I haven't messed with your code. Maybe I have things installed wrong on my system?

**Reply by gpstar81** [reply]
Reference this again: 

https://github.com/gpstar81/haslab-proton-pack/blob/main/ARDUINO.md

It looks like the wavtrigger.h is still trying to reference the AltSoftSerial library. You need to make sure altsoft is commented out in the WavTrigger.h, as per the screenshot in the above link under For compiling code on the Arduino Mega:

**Reply by Beelzebot42** [reply]
I have commented it out and the problem persists. I have deleted all references to AltSoftSerial and it still gives me the following errors.

C:\Users\ruper\AppData\Local\Temp\ccvtnWqh.ltrans0.ltrans.o: In function `update':
c:\arduino\libraries\libraries\WAV-Trigger-Arduino-Serial-Library-master/wavTrigger.cpp:67: undefined reference to `AltSoftSerial::available()'
c:\arduino\libraries\libraries\WAV-Trigger-Arduino-Serial-Library-master/wavTrigger.cpp:68: undefined reference to `AltSoftSerial::read()'
C:\Users\ruper\AppData\Local\Temp\ccvtnWqh.ltrans0.ltrans.o: In function `flush':
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:54: undefined reference to `AltSoftSerial::flushOutput()'
C:\Users\ruper\AppData\Local\Temp\ccvtnWqh.ltrans0.ltrans.o: In function `write':
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:53: undefined reference to `AltSoftSerial::writeByte(unsigned char)'
C:\Users\ruper\AppData\Local\Temp\ccvtnWqh.ltrans0.ltrans.o: In function `global constructors keyed to 65535_0_proton_pack.ino.cpp.o.4840':
c:\arduino\libraries\libraries\ezButton\src/ezButton.cpp:34: undefined reference to `vtable for AltSoftSerial'
c:\arduino\libraries\libraries\ezButton\src/ezButton.cpp:34: undefined reference to `vtable for AltSoftSerial'
C:\Users\ruper\AppData\Local\Temp\ccvtnWqh.ltrans0.ltrans.o: In function `AltSoftSerial::begin(unsigned long)':
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:47: undefined reference to `AltSoftSerial::init(unsigned long)'
C:\Users\ruper\AppData\Local\Temp\ccvtnWqh.ltrans0.ltrans.o: In function `flush':
c:\arduino\libraries\libraries\WAV-Trigger-Arduino-Serial-Library-master/wavTrigger.cpp:53: undefined reference to `AltSoftSerial::available()'
c:\arduino\libraries\libraries\WAV-Trigger-Arduino-Serial-Library-master/wavTrigger.cpp:54: undefined reference to `AltSoftSerial::read()'
C:\Users\ruper\AppData\Local\Temp\ccvtnWqh.ltrans0.ltrans.o: In function `AltSoftSerial::~AltSoftSerial()':
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:46: undefined reference to `vtable for AltSoftSerial'
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:46: undefined reference to `vtable for AltSoftSerial'
c:\arduino\libraries\libraries\altsoftserial/altsoftserial.h:46: undefined reference to `AltSoftSerial::end()'
collect2.exe: error: ld returned 1 exit status

exit status 1

Compilation error: exit status 1

**Reply by Beelzebot42** [reply]
It seems the wavTriggerccp is also trying to find it?

**Comment by Beelzebot42** [comment]
This is what I have in my file:

// ==================================================================
// The following defines are used to control which serial class is
//  used. Uncomment only the one you wish to use. If all of them are
//  commented out, the library will use Hardware Serial
//#define __WT_USE_ALTSOFTSERIAL__
//#define __WT_USE_SERIAL1__
//#define __WT_USE_SERIAL2__
#define __WT_USE_SERIAL3__
// ==================================================================

**Reply by gpstar81** [reply]
Do you have the mega selected as your board type in the ide? Are you 100% sure that you are editing the correct wavtrigger.h that the ide is using ?

**Reply by Beelzebot42** [reply]
I do have the correct board. I think I just figured out what I was doing. I had added WavTrigger.h to my sketch. I think i just need to alter the WavTrigger.h file in the WavTrigger library and not add it to the sketch and update? I tried on my wife's computer doing a fresh install and changing the .h file in the library and it compiled without the errors.

BTW thanks for responding and the patience. I figure if I am doing it wrong others will once it goes to public release.

**Reply by Beelzebot42** [reply]
Success! That was it. I suggest clarifying where/how to update WavTrigger.h as someone else will be dumb like me.

**Reply by gpstar81** [reply]
> Success! That was it. I suggest clarifying where/how to update WavTrigger.h as someone else will be dumb like me.

Cool, glad it's working for you now 😃

**Comment by Beelzebot42** [comment]
For anyone following along or having issues. You need to update the WavTrigger.h file that is in the arduino library. (I just used wordpad) following the steps laid out in the setup instructions. Remember that you also will need to
